move enum QgsMapLayerProxyModel::Filter to Qgis

3nids opened the pull request:
this might solve #38472, probably not.

3nids committed:
move enum QgsMapLayerProxyModel::Filter to Qgis

3nids committed:
find and replace

nyalldawson commented on src/core/layertree/qgslayertreefilterproxymodel.h:
```diff
@@ -66,7 +66,7 @@ class CORE_EXPORT QgsLayerTreeFilterProxyModel : public QSortFilterProxyModel
      * If the list is empty, all types are shown.
      */
-    void setFilters( const QgsMapLayerProxyModel::Filters &filters );
+    void setFilters( const Qgis::LayerFilters &filters );
```
```suggestion
    void setFilters( Qgis::LayerFilters filters );
```

nyalldawson commented:
One thing I'm not sure of here (and maybe you know a fix for): when we move flags to enum flags, sip requires a full `setSomething( SomeFlags( Flag1 | Flag2 ))` instead of just `setSomething( Flag1 | Flag2))`

Is this avoidable? It's not great for plugins

3nids committed:
fix imports

3nids committed:
missing declare flag

3nids committed:
do not use scoped enum

3nids committed:
fix

3nids commented:
The issue is the Python type associated with scoped enum, which is Enum.enum while we would need IntFlag.

there is a new annotation in sip 6, /BaseType=IntFlag/ which does this but it's not accessible in sip 4.

So,.I dropped the scope and added a TODO for QGIS 4.

3nids committed:
use scope enum, add API compatibility overloads

3nids committed:
add deprecated

3nids committed:
fix test
